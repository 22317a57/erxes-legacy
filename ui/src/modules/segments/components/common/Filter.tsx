import Button from "modules/common/components/Button";
import { FormControl } from "modules/common/components/form";
import { __ } from "modules/common/utils";
import { operators } from "modules/customers/constants";
import { FlexContent, FlexItem, FlexRightItem } from "modules/layout/styles";
import React from "react";
import { IConditionFilter } from "../../types";

type Props = {
  names: string[];
  filter: IConditionFilter;
  onChange: (filter: IConditionFilter) => void;
  onRemove?: (id: string) => void;
};

type State = {
  key: string;
  currentName: string;
  currentOperator: string;
  currentValue: string;
};

class Filter extends React.Component<Props, State> {
  constructor(props) {
    super(props);

    const { filter } = this.props;

    this.state = {
      key: filter.key || '',
      currentName: filter.name,
      currentOperator: filter.operator,
      currentValue: filter.value,
    };
  }

  onChange = () => {
    const { currentName, currentOperator, currentValue } = this.state;
    const { onChange, filter } = this.props;

    return onChange({
      key: filter.key,
      name: currentName,
      operator: currentOperator,
      value: currentValue,
    });
  }

  onChangeValue = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentValue: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  onChangeNames = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentName: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  onChangeOperators = (e: React.FormEvent<HTMLElement>) => {
    this.setState({ currentOperator: (e.currentTarget as HTMLInputElement).value }, this.onChange);
  }

  renderNames() {
    const { names } = this.props;
    const { currentName } = this.state;

    return (
      <FormControl componentClass="select" placeholder={__("select")} onChange={this.onChangeNames} value={currentName}>
        <option />
        {names.map((name, index) => (
          <option value={name} key={index}>
            {name}
          </option>
        ))}
      </FormControl>
    );
  }

  renderOperators() {
    const { currentOperator } = this.state;

    return (
      <FormControl componentClass="select" placeholder={__("select")} onChange={this.onChangeOperators} value={currentOperator}>
        <option />
        {operators.map(c => (
          <option value={c.value} key={c.value}>
            {c.name}
          </option>
        ))}
      </FormControl>
    );
  }

  onRemove = () => {
    const { onRemove } = this.props;

    if (onRemove) {
      onRemove(this.props.filter.key || '');
    }
  };

  renderRemoveButton = () => {
    const { onRemove } = this.props;

    if (!onRemove) {
      return;
    }

    return <Button btnStyle="danger" size="small" icon="cancel-1" onClick={this.onRemove} />;
  }

  render() {
    const { currentValue } = this.state;

    return (
      <FlexContent>
        <FlexItem>
          {this.renderNames()}
          {this.renderOperators()}
          <input value={currentValue} onChange={this.onChangeValue} />
        </FlexItem>
        <FlexRightItem>
          {this.renderRemoveButton()}
        </FlexRightItem>
      </FlexContent>
    );
  }
}

export default Filter;
