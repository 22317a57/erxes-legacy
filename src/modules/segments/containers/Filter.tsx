import gql from 'graphql-tag';
import { IRouterProps } from 'modules/common/types';
import { router, withProps } from 'modules/common/utils';
import { getConfig, setConfig } from 'modules/inbox/utils';
import React from 'react';
import { compose, graphql } from 'react-apollo';
import { withRouter } from 'react-router';
import Filter from '../components/Filter';
import { queries } from '../graphql';
import { Counts, SegmentsQueryResponse } from '../types';

type Props = {
  contentType: string;
  counts: Counts;
};

type FinalProps = {
  segmentsQuery: SegmentsQueryResponse;
} & Props &
  IRouterProps;

const STORAGE_KEY = `erxes_sidebar_section_config`;

const FilterContainer = (props: FinalProps) => {
  const { segmentsQuery, history } = props;

  const currentSegment = router.getParam(history, 'segment');

  const toggleSection = ({
    name,
    isOpen
  }: {
    name: string;
    isOpen: boolean;
  }) => {
    // const customerId = this.props.conversation.customerId;
    const config = getConfig(STORAGE_KEY);

    config[name] = isOpen;

    setConfig(STORAGE_KEY, config);

    // this.getCustomerDetail(customerId);
  };

  const setSegment = segment => {
    router.setParams(history, { segment });
  };

  const removeSegment = () => {
    router.removeParams(history, 'segment');
  };

  const extendedProps = {
    ...props,
    currentSegment,
    setSegment,
    removeSegment,
    segments: segmentsQuery.segments || [],
    loading: segmentsQuery.loading,
    toggleSection,
    config: getConfig(STORAGE_KEY)
  };

  return <Filter {...extendedProps} />;
};

export default withProps<Props>(
  compose(
    graphql(gql(queries.segments), {
      name: 'segmentsQuery',
      options: ({ contentType }: { contentType: string }) => ({
        variables: { contentType }
      })
    })
  )(withRouter<FinalProps>(FilterContainer))
);
