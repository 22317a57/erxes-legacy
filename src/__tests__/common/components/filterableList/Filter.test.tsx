import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Filter from '../../../../modules/common/components/filterableList/Filter';

describe('Filter component', () => {
  const defaultProps = {
    onChange: (e: React.FormEvent<HTMLElement>) => null
  };

  test('renders Filter successfully', () => {
    shallow(<Filter {...defaultProps} />);
  });

  test('renders successfully with default value', () => {
    const wrapper = mount(<Filter {...defaultProps} />);
    const props = wrapper.props();

    expect(props).toMatchObject(defaultProps);
  });

  test('snapshot matches', () => {
    const rendered = renderer.create(<Filter {...defaultProps} />).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
