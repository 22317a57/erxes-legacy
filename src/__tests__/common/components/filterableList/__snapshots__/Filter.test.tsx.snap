// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component snapshot matches 1`] = `
<input
  autoFocus={true}
  className="sc-kaNhvL bKDVoH"
  disabled={false}
  onChange={[Function]}
  placeholder="Search"
  required={false}
  type="text"
/>
`;
